feat: Add ECharts Timeseries plugin

villebro opened the pull request:
🏆 Enhancements
Add two new ECharts viz plugins: Timeseries and Pie (WIP).

### SCREENSHOTS
Regular line chart
![image](https://user-images.githubusercontent.com/33317356/90490788-b98a5600-e147-11ea-87c4-da6dee86d69e.png)
Area Chart
![image](https://user-images.githubusercontent.com/33317356/90490634-81831300-e147-11ea-97df-440f6e7b874a.png)
Same chart with 52 week predictions (requires installation of `fbprophet` on backend)
![image](https://user-images.githubusercontent.com/33317356/90489989-be9ad580-e146-11ea-8f3f-439dfafdb4f2.png)
Same predictive chart as stacked area chart
![image](https://user-images.githubusercontent.com/33317356/90490120-e4c07580-e146-11ea-8fce-ed41ef6cdfa7.png)
Storybook entry:
![image](https://user-images.githubusercontent.com/33317356/91708470-0a5b6f00-eb8a-11ea-9abf-fdb7e8a164e8.png)

mistercrunch commented:
Screenshots plz! :)

mistercrunch commented:
![heart](https://user-images.githubusercontent.com/487433/90592631-e4ea6080-e19a-11ea-9424-380a6534d72b.gif)

rusackas commented:
Seems like it might be a good idea to add Storybook entries as well.

ktmud commented on plugins/plugin-chart-echarts/src/Pie/EchartsPie.tsx:
```diff
@@ -0,0 +1,78 @@
+    myChart.resize({
+      width,
+      height,
+    });
```
I believe you can pass `width` and `height` to `init(...)`: https://echarts.apache.org/en/api.html#echarts.init

villebro replied:
The updated `width` and `height` still need to be passed to `resize()`, as otherwise it will get stuck in the original dimensions when resized. But I'll add them to `opts` in `init()` for clarity (perhaps it will make initial rendering faster when no resizing happens). Might also make sense to break out `init()` into an only once runnable `useEffect`.

ktmud commented on plugins/plugin-chart-echarts/src/Pie/controlPanel.ts:
```diff
@@ -0,0 +1,149 @@
+   * - validateNonEmpty: must have at least one value
+   * - validateInteger: must be an integer value
+   * - validateNumber: must be an intger or decimal value
+   */
```
Clean up the comments?

rusackas replied:
Agreed... a lot of the files have some level of generator "instructions" that we can strip out to make this thing ready for primetime.

ktmud commented on plugins/plugin-chart-echarts/src/Timeseries/EchartsTimeseries.tsx:
```diff
@@ -0,0 +1,51 @@
+  width: ${({ width }) => width};
+`;
+
+export default function EchartsTimeseries(props: EchartsTimeseriesProps) {
```
Can we DRY and reuse the same rendering component for all chart types?

villebro replied:
Good idea

ktmud commented on plugins/plugin-chart-word-cloud/src/chart/WordCloud.tsx:
```diff
@@ -132,6 +132,8 @@ class WordCloud extends React.PureComponent<
     const { width, height, encoding } = this.props;
     const { words } = this.state;
+    console.log('state', this.state);
+    console.log('props', this.props);
```
cleanup

rusackas commented on plugins/plugin-chart-echarts/README.md:
````diff
@@ -0,0 +1,54 @@
+/>
+```
+
+### File structure generated
````
This stuff could probably be killed off... it's more for the plugin author, but now we should rewrite the readme for the plugin consumer.

rusackas commented on plugins/plugin-chart-echarts/src/Pie/buildQuery.ts:
```diff
@@ -0,0 +1,41 @@
+ */
+import { buildQueryContext, QueryFormData } from '@superset-ui/query';
+
+/**
```
These instructions can probably get stripped out too.

rusackas commented on plugins/plugin-chart-echarts/src/Pie/transformProps.ts:
```diff
@@ -0,0 +1,107 @@
+    },
+    series: [
+      {
+        // name: 'gender',
```
old prototype comment to delete

rusackas commented on plugins/plugin-chart-echarts/src/Pie/types.ts:
```diff
@@ -0,0 +1,15 @@
+import { DataRecord, ChartProps } from '@superset-ui/chart';
```
Needs Apache boilerplate comments

rusackas replied:
```suggestion
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { DataRecord, ChartProps } from '@superset-ui/chart';
```

rusackas commented on plugins/plugin-chart-echarts/src/Timeseries/transformProps.ts:
```diff
@@ -0,0 +1,203 @@
+    tooltip: {
+      trigger: 'axis',
+      formatter: (params, ticket, callback) => {
+        // @ts-ignore
```
There's a handful of scattered `ts-ignore`s still... I'm not losing any sleep, but maybe we can take a pass at them and see how painful they are to strip out.

villebro replied:
These were actually quite painful, as the current `@types` definitions aren't always coherent. I'd defer the bulk of these unil ECharts 5.0 comes out, as it has way better typing built in to the library than the current 4.x branch.

rusackas commented on plugins/plugin-chart-echarts/src/Timeseries/types.ts:
```diff
@@ -0,0 +1,42 @@
+import { DataRecord, DataRecordValue } from '@superset-ui/chart';
```
Needs Apache license comments

rusackas commented on plugins/plugin-chart-echarts/test/Timeseries/transformProps.test.ts:
```diff
@@ -0,0 +1,38 @@
+import 'babel-polyfill';
```
Needs Apache stuff

rusackas commented on plugins/plugin-chart-echarts/test/index.test.ts:
```diff
@@ -0,0 +1,34 @@
+import { EchartsPieChartPlugin, EchartsTimeseriesChartPlugin } from '../src';
+
+/**
+ * The example tests in this file act as a starting point, and
```
Can nix these comments too

villebro commented:
@ktmud @rusackas this should be ready for a second round of review. Apart from trivial changes like adding license headers, I did the following:
- introduced a `useEchartsComponent` hook that hides the `useEffect` boilerplate.
- did some small tuning to types. These are probably not quite there yet, but I think we can live with these for now (I think we need to do a few small tweaks in the core library, too).
- The Pie chart needs some additional work before being ready for prime time. I added todos for them in `transformProps.ts`: Most of them are easy to fix, but I don't have time to fix them just yet.
So I propose making `Timeseries` usable and fix `Pie` after that.

ktmud commented on plugins/plugin-chart-echarts/src/Timeseries/transformProps.ts:
```diff
@@ -0,0 +1,203 @@
+    const isConfidenceBand = [
+      ForecastSeriesEnum.ForecastLower,
+      ForecastSeriesEnum.ForecastUpper,
+    ].includes(forecastSeries.type);
```
Nit: if it's only two options, I personally would just write 

```ts
const isConfidenceBand =
  forecastSeries.type === ForecastSeriesEnum.ForecastLower ||
  forecastSeries.type === ForecastSeriesEnum.ForecastUpper;
```

villebro commented:
@ktmud this should be ready for another round of review. I redid the `useEffect`s to a full `<Echart>` component to encapsulate all `useEffect`s to minimize rerendering overhead. I'm not sure if it made it noticeably faster, but it feels like it rerenders slightly quicker than before. Check https://github.com/apache-superset/superset-ui/pull/737/commits/45de54b471a2676c446c667df53146e415c77513 for the component commit. In addition I addressed the other comments in https://github.com/apache-superset/superset-ui/pull/737/commits/88a67d48a1f440a4d7c14985c1eaac90ca2253a4 .

rusackas commented on plugins/plugin-chart-echarts/README.md:
````diff
@@ -0,0 +1,40 @@
+  .register();
+```
+
+Then use it via `SuperChart`. See [storybook](https://apache-superset.github.io/superset-ui/?selectedKind=plugin-chart-echarts-line) for more details.
````
I don't _think_ this story exists. We should either remove this line, or add the story.

villebro replied:
Story has now been added with all applicable knobs (check screenshot)

rusackas commented:
Just took a look at @ktmud's feedback, and his suggestions look great. I'll leave the approval because... well... it works! But I agree (and am grateful for) his suggestions.

villebro commented:
Grr, notifications for these review comments had again gone into my spam folder. I'll address the comments and add the storybook entry by Friday morning PST!

villebro commented on packages/superset-ui-demo/storybook/stories/plugins/plugin-chart-echarts/Stories.tsx:
```diff
@@ -0,0 +1,60 @@
+};
+
+export const Timeseries = ({ width, height }) => {
+  const enableForecast = radios('Enable forecast', { Yes: 'yes', No: 'no' }, 'yes');
```
For some reason `radios` only seems to support string values.

rusackas replied:
That's just the nature of radio buttons, as form data inputs, usually used for non-boolean options. You can use the `boolean` knob instead of `radios`, which displays a checkbox, if that's easier/better for you.

villebro commented on packages/superset-ui-demo/storybook/stories/plugins/plugin-chart-echarts/Stories.tsx:
```diff
@@ -0,0 +1,60 @@
+export const Timeseries = ({ width, height }) => {
+  const enableForecast = radios('Enable forecast', { Yes: 'yes', No: 'no' }, 'yes');
+  const forecastEnabled = enableForecast === 'yes';
+  const queryData = data
+    .map(row =>
+      forecastEnabled
+        ? row
+        : {
+            __timestamp: row.__timestamp,
+            Boston: row.Boston,
+            California: row.California,
+            WestTexNewMexico: row.WestTexNewMexico,
+          },
+    )
+    .filter(row => forecastEnabled || !!row.Boston);
```
Here we simulate what the backend would do, i.e. only returning observations if forecasts aren't enabled.

villebro commented:
@ktmud @rusackas I've now introduced a storybook and did those fixes Jesse proposed (thanks so much for patiently helping me with this!). I look forward to improving this going forward, so if we can live with this as the version I'd be stoked if we can press merge and introduce this in `incubator-superset` asap!

rusackas commented on packages/superset-ui-demo/storybook/stories/plugins/plugin-chart-echarts/Stories.tsx:
```diff
@@ -0,0 +1,60 @@
+        ),
+        logAxis: radios('Log axis', { Yes: 'yes', No: 'no' }, 'no') === 'yes',
+        yAxisFormat: 'SMART_NUMBER',
+        stack: radios('Stack', { Yes: 'yes', No: 'no' }, 'no') === 'yes',
```
```suggestion
        stack: boolean('Stack', false),
```
^^ example boolean knob

villebro replied:
Uhh, thanks, I totally missed that!

kristw commented on plugins/plugin-chart-echarts/src/Pie/EchartsPie.tsx:
```diff
@@ -0,0 +1,25 @@
+import { EchartsPieProps } from './types';
+import Echart from '../components/Echart';
+
+export default function EchartsPie(props: EchartsPieProps) {
```
better use destructuring.

kristw commented on plugins/plugin-chart-echarts/src/Pie/buildQuery.ts:
```diff
@@ -0,0 +1,27 @@
+import { buildQueryContext, QueryFormData } from '@superset-ui/query';
+
+export default function buildQuery(formData: QueryFormData) {
+  return buildQueryContext(formData, baseQueryObject => [
```
The second parameter of `buildQueryContext` is not needed in this case.. This is same with the default behavior.

kristw commented on plugins/plugin-chart-echarts/src/Pie/index.ts:
```diff
@@ -0,0 +1,52 @@
+
+const metadata = new ChartMetadata({
+  description: 'Pie chart using ECharts',
+  name: t('EchartsPie'),
```
by creating the metadata here instead of inside the constructor of the plugin, you need to ensure the translator is initialized **before importing this file**. 

Better create the metadata inside the constructor of the plugin

villebro replied:
It appears this may be a problem in several plugins, as most (all?) seem to define metadata outside the constructor. I'll leave this as-if for now and clean up all plugins in one sweep (we also need to fix the yeoman generator).

kristw replied:
I have added the fix in the newer plugins. 
https://github.com/apache-superset/superset-ui/blob/master/plugins/preset-chart-xy/src/BoxPlot/index.ts

kristw commented on plugins/plugin-chart-echarts/src/Pie/transformProps.ts:
```diff
@@ -0,0 +1,87 @@
+  const keys = Array.from(new Set(data.map(datum => datum[groupby[0]])));
+
+  const transformedData = data.map(datum => {
+    return {
```
`return` is not needed.

```ts
data.map(datum => ({ value: ..., name: ... });
```

What if `groupby.length === 0` or `metrics.length===0`, will this crash?

villebro replied:
FYI: Pie is still WIP, so this needs to be cleaned up on many fronts.

kristw commented on plugins/plugin-chart-echarts/src/Timeseries/buildQuery.ts:
```diff
@@ -0,0 +1,88 @@
+                .map(metric => metric.label)
+                .reduce(
+                  (obj, cur) => ({
+                    ...obj,
```
spreading the same object again and again is not efficient.
since `obj` is not used anywhere else it is better to mutate here.

kristw commented on plugins/plugin-chart-echarts/src/Pie/index.ts:
```diff
@@ -0,0 +1,52 @@
+    super({
+      buildQuery,
+      controlPanel,
+      loadChart: () => import('./EchartsPie'),
```
You could 

```ts
import('../components/Echart')
```` 

here instead. No need to create `EChartsPie.tsx`

villebro replied:
Good idea. For now the current logic seems to be a common pattern throughout the existing plugins, so I think it makes sense to tackle this in a separate PR also.

ktmud commented on plugins/plugin-chart-echarts/src/Timeseries/buildQuery.ts:
```diff
@@ -0,0 +1,88 @@
+              columns: formData.groupby,
+              // Create 'dummy' sum aggregates to assign cell values in pivot table
+              aggregates: baseQueryMetrics
+                .map(metric => metric.label)
```
Try this:

```ts
Object.fromEntries(baseQueryMetrics.map(
  metric => [metric.label, { operator: 'sum' }]
))
```

villebro replied:
Much better, thanks!

villebro commented:
Big thanks @ktmud @rusackas @kristw for all your help getting this first ECharts based plugin ready! I'm going to merge this now and look forward to following up with follow-up PRs to address some of the out of scope comments.
